sdk/src/wallet/migration/chrysalis.rs

```diff
@@ -0,0 +1,310 @@
+
+use std::{
+    collections::{HashMap, HashSet},
+    convert::TryInto,
+    io::Read,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use crypto::{
+    ciphers::{chacha::XChaCha20Poly1305, traits::Aead},
+    macs::hmac::HMAC_SHA512,
+};
+use rocksdb::{IteratorMode, Options, DB};
+use serde::Serialize;
+use serde_json::Value;
+use zeroize::Zeroizing;
+
+use crate::{
+    client::{constants::IOTA_COIN_TYPE, Password},
+    types::block::address::Bech32Address,
+    wallet::{
+        migration::{MigrationData, MIGRATION_VERSION_KEY},
+        storage::constants::{
+            ACCOUNTS_INDEXATION_KEY, ACCOUNT_INDEXATION_KEY, CHRYSALIS_STORAGE_KEY, SECRET_MANAGER_KEY,
+            WALLET_INDEXATION_KEY,
+        },
+        Error, Result,
+    },
+};
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AccountAddress {
+    address: Bech32Address,
+    key_index: u32,
+    internal: bool,
+    used: bool,
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct AccountDetailsDto {
+    pub(crate) index: u32,
+    coin_type: u32,
+    alias: String,
+    public_addresses: Vec<AccountAddress>,
+    internal_addresses: Vec<AccountAddress>,
+    addresses_with_unspent_outputs: Vec<Value>,
+    outputs: HashMap<String, Value>,
+    locked_outputs: HashSet<String>,
+    unspent_outputs: HashMap<String, Value>,
+    transactions: HashMap<String, Value>,
+    pending_transactions: HashSet<String>,
+    incoming_transactions: HashMap<String, Value>,
+    native_token_foundries: HashMap<String, Value>,
+}
+
+pub async fn migrate_db_chrysalis_to_stardust(
+    storage_path: impl Into<PathBuf> + Send,
+    password: Option<Password>,
+) -> Result<()> {
+    let storage_path_string = storage_path.into();
+    // `/db` will be appended to the chrysalis storage path, because that's how it was done in the chrysalis wallet
+    let chrysalis_storage_path = &(*storage_path_string).join("db");
+
+    let chrysalis_data = get_chrysalis_data(chrysalis_storage_path, password)?;
+
+    // create new accounts base on previous data
+    let (new_accounts, secret_manager_dto) = migrate_from_chrysalis_data(&chrysalis_data, &storage_path_string, false)?;
+
+    // convert to string keys
+    let chrysalis_data_with_string_keys = chrysalis_data
+        .into_iter()
+        .map(|(k, v)| {
+            Ok((
+                String::from_utf8(k).map_err(|_| Error::Migration("invalid utf8".into()))?,
+                v,
+            ))
+        })
+        .collect::<Result<HashMap<String, String>>>()?;
+
+    log::debug!(
+        "Chrysalis data: {}",
+        serde_json::to_string_pretty(&chrysalis_data_with_string_keys)?
+    );
+
+    let mut opts = Options::default();
+    opts.create_if_missing(true);
+    opts.create_missing_column_families(true);
+    let stardust_db = DB::open(&opts, storage_path_string).unwrap();
+
+    // store chrysalis data in a new key
+    stardust_db.put(
```

Comment:
What about encryption? Should we also encrypt the stardust db if a password is given?